source/fuzz/transformation_propagate_instruction_down.h

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2020 Vasyl Teliman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_FUZZ_TRANSFORMATION_PROPAGATE_INSTRUCTION_DOWN_H_
+#define SOURCE_FUZZ_TRANSFORMATION_PROPAGATE_INSTRUCTION_DOWN_H_
+
+#include <map>
+
+#include "source/fuzz/protobufs/spirvfuzz_protobufs.h"
+#include "source/fuzz/transformation.h"
+#include "source/fuzz/transformation_context.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace fuzz {
+
+class TransformationPropagateInstructionDown : public Transformation {
+ public:
+  explicit TransformationPropagateInstructionDown(
+      const protobufs::TransformationPropagateInstructionDown& message);
+
+  TransformationPropagateInstructionDown(
+      uint32_t block_id, uint32_t phi_fresh_id,
+      const std::map<uint32_t, uint32_t>& successor_id_to_fresh_id);
+
+  // - it should be possible to apply this transformation to |block_id| (see
+  //   IsApplicableToBlock method).
+  // - every acceptable successor of |block_id| (see GetAcceptableSuccessors
+  //   method)
+  //   must have an entry in the |successor_id_to_fresh_id| map.
+  // - all values in |successor_id_to_fresh_id| and |phi_fresh_id| must be
+  //   unique and fresh.
+  bool IsApplicable(
+      opt::IRContext* ir_context,
+      const TransformationContext& transformation_context) const override;
+
+  // - Adds a clone of the propagated instruction into every acceptable
+  //   successor of |block_id|.
+  // - Removes the original instruction.
+  // - Creates an OpPhi instruction if possible, that tries to group created
+  //   clones.
+  // - If the original instruction's id was irrelevant - marks created
+  //   instructions as irrelevant. Otherwise, marks the created instructions as
+  //   synonymous to each other.
+  void Apply(opt::IRContext* ir_context,
+             TransformationContext* transformation_context) const override;
+
+  protobufs::Transformation ToMessage() const override;
+
+  // Returns true if this transformation can be applied to the block with id
+  // |block_id|. Concretely, returns true iff:
+  // - |block_id| must be a result id of some reachable basic block in the
+  //   module.
```

Comment:
Grammar: after "returns true iff", you should use "is" or "has" rather than "must be" or "must have", etc.

Reply:
Fixed in c4faca9.